src/backend.rs

```diff
@@ -333,93 +727,253 @@ pub fn br(ctx: &mut Context, label: Label) {
+    let mut out = ArrayVec::new();
 
-    let label = &ctx.func_defs[index as usize].label;
-    dynasm!(ctx.asm
-        ; call =>*label
-    );
+    // TODO: If there are no `StackValue::Pop`s that need to be popped
+    //       before we reach our `Temp` value, we can set the `StackValue`
+    //       for the register to be restored to `StackValue::Pop` (and
+    //       release the register!) instead of restoring it.
+    for &reg in SCRATCH_REGS.iter() {
+        if !ctx.block_state.regs.is_free(reg) {
+            dynasm!(ctx.asm
+                ; push Rq(reg)
+            );
+            out.push(reg);
+        }
+    }
+
+    out
+}
+
+/// Write the arguments to the callee to the registers and the stack using the SystemV
+/// calling convention.
+fn pass_outgoing_args(ctx: &mut Context, arity: u32) -> CallCleanup {
+    let num_stack_args = (arity as usize).saturating_sub(ARGS_IN_GPRS.len()) as i32;
+
+    // We pop stack arguments first - arguments are RTL
+    if num_stack_args > 0 {
+        let size = num_stack_args * WORD_SIZE as i32;
 
-    if return_arity == 1 {
+        // Reserve space for the outgoing stack arguments (so we don't
+        // stomp on any locals or the value stack).
         dynasm!(ctx.asm
-            ; push rax
+            ; sub rsp, size
         );
-        ctx.sp_depth.reserve(1);
+        ctx.block_state.depth.reserve(num_stack_args as u32);
+
+        for stack_slot in (0..num_stack_args).rev() {
+            // Since the stack offset is from the bottom of the locals
+            // and we want to start from the actual RSP (so `offset = 0`
+            // writes to `[rsp]`), we subtract our current depth.
+            //
+            // We might want to do this in the future by having a separate
+            // `AbsoluteValueLocation` and `RelativeValueLocation`.
+            let offset =
+                stack_slot * WORD_SIZE as i32 - ctx.block_state.depth.0 as i32 * WORD_SIZE as i32;
+            pop_i32_into(ctx, ValueLocation::Stack(offset));
+        }
+    }
+
+    for reg in ARGS_IN_GPRS[..(arity as usize).min(ARGS_IN_GPRS.len())]
+        .iter()
+        .rev()
+    {
+        pop_i32_into(ctx, ValueLocation::Reg(*reg));
+    }
+
+    CallCleanup {
+        stack_depth: num_stack_args,
+        restore_registers: save_volatile(ctx),
     }
 }
 
-pub fn prologue(ctx: &mut Context, stack_slots: u32) {
+/// Frees up the stack space used for stack-passed arguments and restores the value
+/// of volatile (i.e. caller-saved) registers to the state that they were in before
+/// the call.
+fn post_call_cleanup(ctx: &mut Context, mut cleanup: CallCleanup) {
+    if cleanup.stack_depth > 0 {
+        let size = cleanup.stack_depth * WORD_SIZE as i32;
+        dynasm!(ctx.asm
+            ; add rsp, size
+        );
+    }
+
+    for reg in cleanup.restore_registers.drain(..).rev() {
+        dynasm!(ctx.asm
+            ; pop Rq(reg)
+        );
```

Comment:
I see there's a `TODO` on `set_local_i32` to put locals back into registers. I think we can add a similar TODO here too: instead of eagerly restoring all the values after a call, just make `save_volatile` update the bookkeeping to say that they're on the stack, so that we don't need to do anything here, and they'll be reloaded on demand.

Reply:
Excellent point. I want to remove all` push`/`pop` calls and have real stack allocation (with reusing stack slots and lazily increasing stack space in powers of 2), which is required to efficiently do some stuff including using callee-saved registers/unused argument registers as scratch, directly accessing locals in instructions without having to do a linear probe to invalidate them on `set_local`, and skipping prologue/epilogue in functions that never use the physical stack (atm this can only be done for zero-argument functions since we currently always reserve space to save argument registers).